In [55]:
# Needed libraries
import pandas as pd

In [56]:
# Here, we start with our EDA process

# First, we charge the dataset champions_league_2011_2025.csv
df = pd.read_csv("../champions_league_2011_2025.csv")

In [57]:
# First visualizations of df
df.head()

,season,fase,instancia,fecha,hora_inicio,hora_fin,local,visitante,marcador,global,...,amarillas,rojas,posesion_local,posesion_visitante,faltas_total,faltas_local,faltas_visitante,corners_total,corners_local,corners_visitante
0,2010-2011,Octavos,Ida,15-02-2011,20:45,22:45,Valencia,Schalke 04,1-1,2-4,...,Matip 16'; Ricardo Costa 62'; Schmitz 65'; Neu...,Schmitz,65%,35%,30.0,9.0,21.0,12.0,5.0,7.0
1,2010-2011,Octavos,Vuelta,09-03-2011,NaN,NaN,Schalke 04,Valencia,3-1,4-2,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2010-2011,Octavos,Ida,16-02-2011,NaN,NaN,AC Milan,Tottenham,0-1,0-1,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2010-2011,Octavos,Vuelta,09-03-2011,NaN,NaN,Tottenham,AC Milan,0-0,1-0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2010-2011,Octavos,Ida,15-02-2011,NaN,NaN,Arsenal,Barcelona,2-1,3-4,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [58]:
# Fast view of the dataset columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445 entries, 0 to 444
Data columns (total 38 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   season                   445 non-null    object 
 1   fase                     445 non-null    object 
 2   instancia                445 non-null    object 
 3   fecha                    445 non-null    object 
 4   hora_inicio              46 non-null     object 
 5   hora_fin                 46 non-null     object 
 6   local                    445 non-null    object 
 7   visitante                445 non-null    object 
 8   marcador                 445 non-null    object 
 9   global                   409 non-null    object 
 10  estadio                  445 non-null    object 
 11  ciudad                   445 non-null    object 
 12  pais                     445 non-null    object 
 13  arbitro_principal        14 non-null     object 
 14  arbitros_linea           2

In [59]:
# Shape and size of dataset
df.shape

(445, 38)

In [60]:
# View of columns names
df.columns

Index(['season', 'fase', 'instancia', 'fecha', 'hora_inicio', 'hora_fin',
       'local', 'visitante', 'marcador', 'global', 'estadio', 'ciudad', 'pais',
       'arbitro_principal', 'arbitros_linea', 'entrenador_local',
       'entrenador_visitante', 'planteles', 'puntuaciones_jugadores',
       'tiros_totales', 'tiros_totales_local', 'tiros_totales_visitante',
       'tiros_puerta', 'tiros_puerta_local', 'tiros_puerta_visitante', 'goles',
       'asistencias', 'cambios', 'amarillas', 'rojas', 'posesion_local',
       'posesion_visitante', 'faltas_total', 'faltas_local',
       'faltas_visitante', 'corners_total', 'corners_local',
       'corners_visitante'],
      dtype='object')

In [61]:
# Let's view possible null values
df.isnull().sum()

season                       0
fase                         0
instancia                    0
fecha                        0
hora_inicio                399
hora_fin                   399
local                        0
visitante                    0
marcador                     0
global                      36
estadio                      0
ciudad                       0
pais                         0
arbitro_principal          431
arbitros_linea             443
entrenador_local            29
entrenador_visitante        29
planteles                  438
puntuaciones_jugadores     445
tiros_totales              444
tiros_totales_local        444
tiros_totales_visitante    444
tiros_puerta               444
tiros_puerta_local         444
tiros_puerta_visitante     444
goles                       26
asistencias                444
cambios                    443
amarillas                  441
rojas                      431
posesion_local             443
posesion_visitante         443
faltas_t

In [64]:
# Calcule missingness
missingness = df.isnull().mean()
print("Missingness:", missingness)

Missingness: season                     0.000000
fase                       0.000000
instancia                  0.000000
fecha                      0.000000
hora_inicio                0.896629
hora_fin                   0.896629
local                      0.000000
visitante                  0.000000
marcador                   0.000000
global                     0.080899
estadio                    0.000000
ciudad                     0.000000
pais                       0.000000
arbitro_principal          0.968539
arbitros_linea             0.995506
entrenador_local           0.065169
entrenador_visitante       0.065169
planteles                  0.984270
puntuaciones_jugadores     1.000000
tiros_totales              0.997753
tiros_totales_local        0.997753
tiros_totales_visitante    0.997753
tiros_puerta               0.997753
tiros_puerta_local         0.997753
tiros_puerta_visitante     0.997753
goles                      0.058427
asistencias                0.997753
cambios        

In [65]:
# Sparsity or percent of empty cells in dataset
total_celdas = df.size
celdas_vacias = df.isnull().sum().sum()

sparsity = celdas_vacias / total_celdas

print(f"Sparsity: {sparsity:.2%}")

Sparsity: 62.64%


We notice there are several null values on some clumns, such as arbitro_principal, tiros_totales o tiros_puerta_visitante. Some are very stadistic information and data, exact and to the letter. Sometimes, difficult to find all that data.

We have to adjust the size of the dataset according to the columns that can be deleted, not important information. For example, columns like hora_inicio and hora_fin seem to be dispensable

445 rows, many columnas with 444 or 445 rows with nulll values on specific columns. That would not be useful for us in our analysis.

However, there are 445 rows in our dataset. Deleting hora_inicio and hora_fin and their rows would be the opposite of a good and severe analysis and recommendation. So, in order to prevent that type of troubles, we delete only the columnas, as we consider before, dispensable.

In [32]:
null_ratio = df.isnull().mean().sort_values(ascending=False)
# Null values among total rows of dataset
# 0-1
print(null_ratio)

puntuaciones_jugadores     1.000000
tiros_totales              0.997753
asistencias                0.997753
tiros_puerta_visitante     0.997753
tiros_puerta_local         0.997753
tiros_puerta               0.997753
tiros_totales_visitante    0.997753
tiros_totales_local        0.997753
corners_local              0.995506
faltas_visitante           0.995506
faltas_local               0.995506
posesion_visitante         0.995506
posesion_local             0.995506
cambios                    0.995506
corners_visitante          0.995506
arbitros_linea             0.995506
amarillas                  0.991011
planteles                  0.984270
arbitro_principal          0.968539
rojas                      0.968539
corners_total              0.939326
faltas_total               0.939326
hora_fin                   0.896629
hora_inicio                0.896629
global                     0.080899
entrenador_local           0.065169
entrenador_visitante       0.065169
goles                      0

In [33]:
df_copy = df.copy()

In [34]:
# We will skip and delete those columns with more than 90% NULL on their values
cols_to_drop = null_ratio[null_ratio > 0.9].index
df_copy = df_copy.drop(columns=cols_to_drop)

In [36]:
df_copy.columns

Index(['season', 'fase', 'instancia', 'fecha', 'hora_inicio', 'hora_fin',
       'local', 'visitante', 'marcador', 'global', 'estadio', 'ciudad', 'pais',
       'entrenador_local', 'entrenador_visitante', 'goles'],
      dtype='object')

In [37]:
# For practical issues, we will delete hora_inicio and hora_fin
horas =["hora_inicio", "hora_fin"]
df_copy = df_copy.drop(columns=horas)

In [38]:
df_copy.columns

Index(['season', 'fase', 'instancia', 'fecha', 'local', 'visitante',
       'marcador', 'global', 'estadio', 'ciudad', 'pais', 'entrenador_local',
       'entrenador_visitante', 'goles'],
      dtype='object')

In [40]:
df_copy.shape

(445, 14)

In [ ]:
# Verify duplicated 
filas_repetidas = df_copy[df_copy.duplicated(keep=False)]
print(filas_repetidas.sort_values(by=list(df_copy.columns)))

Empty DataFrame
Columns: [season, fase, instancia, fecha, local, visitante, marcador, global, estadio, ciudad, pais, entrenador_local, entrenador_visitante, goles]
Index: []


There are not duplicated rows. That's nice.

Now, we have to get sure our column fecha is in Date format.

In [46]:
print(df_copy['fecha'].dtype)

object


It is object type. Python reads fecha as a String.

In [47]:
df_copy['fecha'] = pd.to_datetime(df_copy['fecha'])

C:\Users\lcanc\AppData\Local\Temp\ipykernel_21780\3685500320.py:1: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  df_copy['fecha'] = pd.to_datetime(df_copy['fecha'])


In [48]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445 entries, 0 to 444
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   season                445 non-null    object        
 1   fase                  445 non-null    object        
 2   instancia             445 non-null    object        
 3   fecha                 445 non-null    datetime64[ns]
 4   local                 445 non-null    object        
 5   visitante             445 non-null    object        
 6   marcador              445 non-null    object        
 7   global                409 non-null    object        
 8   estadio               445 non-null    object        
 9   ciudad                445 non-null    object        
 10  pais                  445 non-null    object        
 11  entrenador_local      416 non-null    object        
 12  entrenador_visitante  416 non-null    object        
 13  goles               

Now, we are sure fecha is a Date type column of values.

We have explored possible dispensable data. Now, we can clean and produce our new CSV document.

Now, we will create our Data Dictionary, as it orders the instructiones for this first advance of the heavy Big Data project.

In [50]:
# We convert values in column marcador to numeric instead of object type
if "marcador" in df_copy.columns:
        goles_split = df_copy["marcador"].str.extract(r'(\d+)\D+(\d+)')
        df_copy["home_goals"] = pd.to_numeric(goles_split[0], errors="coerce")
        df_copy["away_goals"] = pd.to_numeric(goles_split[1], errors="coerce")

In [51]:
# A better way to manage some null values that did not die after the deleteing of some
df_copy = df_copy.fillna({
        "entrenador_local": "unknown",
        "entrenador_visitante": "unknown",
        "goles": "unknown",
        "global": "not_applicable"
    })

In [52]:
def create_data_dictionary(df: pd.DataFrame) -> pd.DataFrame:
    descriptions = {
        "season": "Season of the competition",
        "fase": "Competition phase (group stage, knockout)",
        "instancia": "Specific round (e.g., quarter-final)",
        "fecha": "Match date",
        "local": "Home team",
        "visitante": "Away team",
        "marcador": "Match score (home-away)",
        "home_goals": "Goals scored by home team",
        "away_goals": "Goals scored by away team",
        "global": "Aggregate score in knockout rounds",
        "estadio": "Stadium name",
        "ciudad": "City of the stadium",
        "pais": "Country where match was played",
        "entrenador_local": "Home team coach",
        "entrenador_visitante": "Away team coach",
        "goles": "Goal details or scorers"
    }

    data_dict = pd.DataFrame({
        "column": df.columns,
        "dtype": df.dtypes.astype(str),
        "missing": df.isnull().sum(),
        "missing_%": (df.isnull().mean() * 100).round(2),
        "description": [descriptions.get(col, "No description") for col in df.columns],
        "example": [str(df[col].dropna().iloc[0]) if df[col].dropna().shape[0] > 0 else "" for col in df.columns]
    })

    return data_dict

In [53]:
print("📚 Generating data dictionary...")
data_dict = create_data_dictionary(df_copy)
data_dict.to_csv("../data/processed/data_dictionary_v1.csv", index=False)

📚 Generating data dictionary...


SCALE ANALYSIS

In [54]:
rows, cols = df_copy.shape
missing = df_copy.isnull().mean()

print("Rows:", rows)
print("Columns:", cols)
print("Missing ratio:\n", missing)

Rows: 445
Columns: 16
Missing ratio:
 season                  0.0
fase                    0.0
instancia               0.0
fecha                   0.0
local                   0.0
visitante               0.0
marcador                0.0
global                  0.0
estadio                 0.0
ciudad                  0.0
pais                    0.0
entrenador_local        0.0
entrenador_visitante    0.0
goles                   0.0
home_goals              0.0
away_goals              0.0
dtype: float64


The dataset includes 445 rows, 16 columnas by now, no missing ratio after cleaning and sustitution of some null values on some columns.